src/systems/filecoin_blockchain/chainsync/_index.md

```diff
@@ -1,58 +1,490 @@
+  - The chain state provided includes the `Blocks` and `StateTree` for the `Finality` epoch, as well as
+    candidate `Blocks` and `StateTrees` for unfinalized epochs.
+- **security conditions to transition out:**
+  - Temporary network partitions (see _Detecting Network Partitions_).
+  - Encounter gaps of `>MaxBlockGap` or `>MaxEpochGap` between Validated set and a new `ChainSync.BestTargetHead`
+- **transitions out:**
+  - if a temporary network partition is detected: move to `CHAIN_CATCHUP`
+  - if `BlockGap > MaxBlockGap`: move to `CHAIN_CATCHUP`
+  - if `EpochGap > MaxEpochGap`: move to `CHAIN_CATCHUP`
+  - if node is shut down: move to `INIT`
+
+# Block Fetching, Validation, and Propagation
+
+## Notes on changing `TargetHeads` while syncing
+
+- `TargetHeads` is changing, as `ChainSync` must be aware of the best heads at any time. reorgs happen, and our first set of peers could've been bad, we keep discovering others.
+  - Hello protocol is good, but it's polling. unless node is constantly polllng, wont see all the heads.
+  - `BlockPubsub` gives us the realtime view into what's actually going on.
+  - weight can also be close between 2+ possible chains (long-forked), and `ChainSync` must select the right one (which, we may not be able to distinguish until validating all the way)
+- fetching + validation are strictly faster per round on average than blocks produced/block time (if they're not, will always fall behind), so we definitely catch up eventually (and even quickly). the last couple rounds can be close ("almost got it, almost got it, there").
+
+## General notes on fetching Blocks
+
+- `ChainSync` selects and maintains a set of the most likely heads to be correct from among those received
+  via `BlockPubsub`. As more blocks are received, the set of `TargetHeads` is reevaluated.
+- `ChainSync` fetches `Blocks`, `Messages`, and `StateTree` through the `Graphsync` protocol.
+- `ChainSync` maintains sets of `Blocks/Tipsets` in `Graphs` (see `ChainSync.id`)
+- `ChainSync` gathers a list of `TargetHeads` from `BlockPubsub`, sorted by likelihood of being the best chain (see below).
+- `ChainSync` makes requests for chains of `BlockHeaders` to close gaps between  `TargetHeads`
+- `ChainSync` forms partial unvalidated chains of `BlockHeaders`, from those received via `BlockPubsub`, and those requested via `Graphsync`.
+- `ChainSync` attempts to form fully connected chains of `BlockHeaders`, parting from `StateTree`, toward observed `Heads`
+- `ChainSync` minimizes resource expenditures to fetch and validate blocks, to protect against DOS attack vectors.
+  `ChainSync` employs **Progressive Block Validation**, validating different facets at different stages of syncing.
+- `ChainSync` delays syncing `Messages` until they are needed. Much of the structure of the partial chains can
+  be checked and used to make syncing decisions without fetching the `Messages`.
+
+## Progressive Block Validation
```

Comment:
I think this section should be split out into its own md. We've had multiple issues in the past with validation being defined in a few spots. Given how key a topic it is, relevance to consensus, etc. I think it probably warrants a standalone block val section that everything else links to.

I should be touching it up wrt consensus as well.

Reply:
Specifically we should have a canonical way of specifying (ie code) how you run various validations (eg ChainEpoch) and having a standalone file in which to link to the code would be great.

Reply:
As an implementer I think it is really critical to keep this the sequencing description in the syncing section.  Sequencing is a key thing that up to now has been woefully underspecified. 
 Validation order matters and is non-trivially sequenced along with fetching, propagating and state updating.

I'm not sure how much of the above suggestion is about spec organization so apologies if I'm misunderstanding.  For example I'm totally happy if Progressive Block Valdation continues to exist with a full description of the sequence but points into those functions in the section you're proposing.

Reply:
> I think this section should be split out into its own md. We've had multiple issues in the past with validation being defined in a few spots. Given how key a topic it is, relevance to consensus, etc. I think it probably warrants a standalone block val section that everything else links to.

I'll probably put the functions themselves into chain package (block validation only makes sense in terms of a chain), but this section will remain here, as these stages are key to understand in chainsync, and _only_ needed split like this here. I wont put all the specifics of how to validate each thing here -- that belongs in the canonical section. This section here can point to the code + functions in the canonical section.

> My bad confused with previous section

with which section? i agreed with your (now crossed out) comment

Reply:
As long as there is a single canonical detailed section dedicated to how block Val is done, I am happy.